operator/cilium_node.go

```diff
@@ -0,0 +1,121 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+
+	"k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+)
+
+var ciliumNodeStore cache.Store
+
+func convertToCiliumNode(obj interface{}) interface{} {
+	cnp, _ := obj.(*v2.CiliumNode)
+	return cnp
+}
+
+func startSynchronizingCiliumNodes() {
+	log.Info("Starting to synchronize CiliumNode custom resources...")
+
+	ciliumNodeStore = cache.NewStore(cache.DeletionHandlingMetaNamespaceKeyFunc)
+	ciliumNodeInformer := informer.NewInformerWithStore(
+		cache.NewListWatchFromClient(ciliumK8sClient.CiliumV2().RESTClient(),
+			"ciliumnodes", v1.NamespaceAll, fields.Everything()),
+		&v2.CiliumNode{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if node, ok := obj.(*v2.CiliumNode); ok {
+					log.Debugf("Received new CiliumNode %+v", node)
+
+					_, err := k8s.Client().CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
+					switch {
+					case k8serrors.IsNotFound(err):
+						log.Warningf("Discovered CiliumNode %s without matching Kubernetes node resource", node.Name)
+						deleteCiliumNode(node.Name)
+						return
+					case err != nil:
+						log.WithError(err).Warningf("Unable to retrieve Kubernetes node %s", node.Name)
+					}
+
+					ciliumNodeUpdated(node)
+				} else {
+					log.Warningf("Unknown CiliumNode object received: %+v", obj)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if node, ok := newObj.(*v2.CiliumNode); ok {
+					log.Debugf("Received updated CiliumNode %+v", node)
+					ciliumNodeUpdated(node)
+				} else {
+					log.Warningf("Unknown CiliumNode object received: %+v", newObj)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if node, ok := obj.(*v2.CiliumNode); ok {
+					log.Debugf("CiliumNode got deleted %+v", node)
+					ciliumNodeDeleted(node.Name)
+				} else {
+					log.Warningf("Unknown CiliumNode object received: %+v", obj)
+				}
+			},
+		},
+		convertToCiliumNode,
+		ciliumNodeStore,
+	)
+
+	go ciliumNodeInformer.Run(wait.NeverStop)
+
+	// Delay start of the CiliumNode garbage collector
+	go func() {
+		time.Sleep(time.Minute * 2)
+		controller.NewManager().UpdateController("remove-stale-cilium-nodes",
+			controller.ControllerParams{
+				RunInterval: time.Minute * 30,
+				DoFunc: func(ctx context.Context) error {
+					client := k8s.Client()
+					for _, nodeName := range nodeManager.GetNames() {
+						_, err := client.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+						switch {
+						case k8serrors.IsNotFound(err):
```

Comment:
The CiliumNode is bound to the lifecycle of the k8s node so it will get deleted as the k8s node gets deleted which happens 5 minutes after the node has been gone.

The garbage collector exists to clean up any CiliumNode for which the owner reference was not possible to attach.

The watcher will delete nodes on startup for which no k8s node is found but this only happens on startup so if the operator keeps running, leftover nodes could accumulate. I've added a comment to document this better.

Reply:
Correction from my previous statement: The combination of k8s node watcher plus the watcher of CiliumNode Add which will delete resources for which no k8s node can be found is sufficient to ensure removal of CIliumNode resources. I've removed the garbage collector.